cpu/rpx0xx/periph/usbdev.c

```diff
@@ -0,0 +1,860 @@
+{
+    assert(ep_num < 16);
+    __IOM uint32_t * reg_map[] = {
+        &USBCTRL_DPRAM->EP0_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP1_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP2_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP3_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP4_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP5_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP6_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP7_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP8_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP9_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP10_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP11_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP12_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP13_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP14_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP15_IN_BUFFER_CONTROL,
+    };
+    return reg_map[ep_num];
+}
+
+/* Read the ep-related bit from a status register. */
+static inline bool _reg_get_flag_for_ep(__IOM uint32_t *reg, usbdev_ep_t *ep)
+{
+    uint32_t val = *reg;
+    return val & 1 << (2 * ep->num + (ep->dir == USB_EP_DIR_IN ? 0 : 1));
+}
+
+/* Clear the ep-related bit from a status register. */
+static inline void _reg_clear_flag_for_ep(__IOM uint32_t *reg, usbdev_ep_t *ep)
+{
+    io_reg_atomic_clear(reg, 
+                        1 << (2 * ep->num + (ep->dir == USB_EP_DIR_IN ? 0 : 1)));
+}
+
+static inline __IOM uint32_t * _get_ep_buf_ctrl_reg(uint8_t ep_num,
+                                                    usb_ep_dir_t dir)
+{
+    if (dir == USB_EP_DIR_IN) {
+        return _get_ep_in_buf_ctrl_reg(ep_num);
+    } else {
+        return _get_ep_out_buf_ctrl_reg(ep_num);
+    }
+}
+
+static inline __IOM uint32_t * _get_ep_ctrl_reg(uint8_t ep_num,
+                                                usb_ep_dir_t dir)
+{
+    if (dir == USB_EP_DIR_IN) {
+        return _get_ep_in_ctrl_reg(ep_num);
+    } else {
+        return _get_ep_out_ctrl_reg(ep_num);
+    }
+}
+
+/** See @io_reg_write_dont_corrupt() */
+static inline void _ep_ctrl_reg_write(uint8_t ep_num, usb_ep_dir_t dir,
+                                      uint32_t value, uint32_t mask)
+{
+    __IOM uint32_t *reg = _get_ep_ctrl_reg(ep_num, dir);
+    io_reg_write_dont_corrupt(reg, value, mask);
+    busy_wait_at_least_cycles(12);
+}
+
+/** See @io_reg_write_dont_corrupt() */
+static inline void _ep_buf_ctrl_reg_write(uint8_t ep_num, usb_ep_dir_t dir, 
+                                          uint32_t value, uint32_t mask)
+{
+    __IOM uint32_t *reg = _get_ep_buf_ctrl_reg(ep_num, dir);
+    io_reg_write_dont_corrupt(reg, value, mask);
+    busy_wait_at_least_cycles(12);
+}
+
+/* Get DPRAM buffer offset. */
+static inline uintptr_t _dpram_offset(uint8_t *buf)
+{
+    return (uintptr_t) buf ^ (uintptr_t) USBCTRL_DPRAM_BASE;
+}
+
+static rpx0xx_usb_ep_t *_get_ep(uint8_t ep_num, usb_ep_dir_t dir)
+{
+    return &_hw_usb_dev.hw_ep[2 * ep_num + (dir == USB_EP_DIR_OUT ? 0 : 1)];
```

Comment:
```suggestion
    return &_hw_usb_dev.hw_ep[2 * ep_num + (dir == USB_EP_DIR_IN ? 0 : 1)];
```
This looks innocent but this changes is really important.